daft/functions/process.py

```diff
@@ -0,0 +1,78 @@
+"""Process Functions."""
+
+from __future__ import annotations
+
+import subprocess
+from typing import Any, Literal
+
+import daft
+from daft.datatype import DataType, DataTypeLike
+from daft.expressions import Expression
+
+
+def run_process(
+    args: Expression | list[Expression | Any],
+    *,
+    shell: bool = False,
+    on_error: Literal["raise", "ignore", "log"] = "log",
+    return_dtype: DataTypeLike = DataType.string(),
+) -> Expression:
+    """Returns an expression that runs an external process (optionally via a shell) and exposes its stdout as a column.
+
+    This helper wraps a Python UDF around ``subprocess.run`` so it can be used inside DataFrame expressions.
+
+    Args:
+        args (Expression | list[Expression | Any]):
+            The command to execute.
+            If ``shell=False`` (default), pass a list of arguments, for example ``["ls", "-a", col("path")]``.
+            If ``shell=True``, pass a single string expression or a list that will be joined, for example ``"echo hello"`` or ``["echo", "hello"]``.
+        shell (bool, default=False):
+            Whether to execute the command via the system shell (equivalent to ``subprocess.run(..., shell=True)``).
+            Using the shell enables pipes and redirection but is more vulnerable to injection. Defaults to ``False``.
+        on_error (Literal["raise", "ignore", "log"], default="log"):
+            Whether to log an error when encountering an error, or log a warning and return a null
+        return_dtype: Desired Daft data type for the result column. Defaults to a UTF-8 string column.
+
+    Returns:
+        Expression: An expression representing the stdout of the process converted to ``return_dtype`` (defaults to a UTF-8 string column).
+
+    Examples:
+        >>> import daft
+        >>> from daft import col
+        >>> from daft.functions import run_process
+        >>> df = daft.from_pydict({"a": ["hello"], "b": ["world"]})
+        >>> # Without shell
+        >>> expr = run_process(["echo", col("a"), col("b")])
+        >>> df = df.select(expr.alias("out"))
+        >>> df.to_pylist()
+        [{'out': 'hello world'}]
+        >>>
+        >>> # With shell and return_dtype=int
+        >>> df = daft.from_pydict({"x": ["hello world"]})
+        >>> expr = run_process(("echo " + col("x") + " | wc -c"), shell=True, return_dtype=int)
+        >>> df = df.select(expr.alias("word_count"))
+        >>> df.to_pylist()
+        [{'word_count': 12}]
+    """
+
+    @daft.func(return_dtype=return_dtype, on_error=on_error)
```

Comment:
To make the conversion from string to the specified `return_dtype` explicit, could we instead have `return_dtype=DataType.str()` here and then do an explicit cast on the expression before returning it? e.g.

```python
return _impl(*expr_args).cast(return_dtype)
```

It might do the same thing but this is just more explicit.